### PR TITLE
Document MDX optimize static logic

### PR DESCRIPTION
**Pull request by bluwy**
## Changes

Followup to https://github.com/withastro/astro/pull/7151#discussion_r1202441822 suggested by @ematipico

This adds internal docs on how the MDX optimization feature work.

After writing this out, I have no idea if it makes things clearer 😅 I'm not the best explainer and it could be polished a bit more.

## Testing

n/a

## Docs

Adds internal documentation for how MDX optimization works.

### CODE REVIEW

File: packages/integrations/mdx/src/README.md
Comment (ematipico) [inline packages/integrations/mdx/src/README.md:62]:
```suggestion
      A: Because we're looking for nodes whose children are static, the node itself doesn't need to be static.
```

File: packages/integrations/mdx/src/README.md
Comment (ematipico) [inline packages/integrations/mdx/src/README.md:64]:
```suggestion
      A: No, but we'll clear that up later in step 3.
```

File: packages/integrations/mdx/src/README.md
Comment (ematipico) [inline packages/integrations/mdx/src/README.md:70]:
```suggestion
      A: Because if the `node` isn't static, that means all its ancestors (`elementStack`) have non-static children, so they couldn't be a subtree root to be optimized anymore.
```

File: packages/integrations/mdx/src/README.md
Comment (ematipico) [inline packages/integrations/mdx/src/README.md:72]:
```suggestion
      A: If we're at non-static node, the `node` should still be considered in `allPossibleElements` as its children could be static.
```

**Review by ematipico — APPROVED** [review]
To be honest, I find it really clear, and it helps **A LOT**!

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:42]: Thanks for the suggestion! I think the "could" here still makes sense as it _is_ uncertain that it will happen, the condition is later explained  at the "if its children are static part".
```suggestion
> NOTE: If one of the nodes in `pre` is MDX, the optimization will not be applied to `pre`, but could be applied to the inner MDX node if its children are static.
```

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:46]: Yes! A lot more concise

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:55]: Yup, the `elementStack` here usually contains nodes that we're uncertain of being a subtree root.

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:67]: That's fair. I figured it was hard to grasp so started with some context instead 😄

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:68]: This was indeed a tough one to phrase. I made it a step 4 because you wouldn't know why we're doing this until you've read step 2 and 3. So the steps here are kinda a "chain of thought" when reading what this algorithm does.

In practice, this happens before step 2 (as phrased by the "This happens before the code in step 2" part), but if I made this step 2 early on, it would be hard to explain what's going on 😅

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:70]:
```suggestion
      A: Because if the `node` isn't static, that means all its ancestors (`elementStack`) have non-static children. So, the ancestors couldn't be a subtree root to be optimized anymore.
```

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:81]:
```suggestion
Astro's MDX implementation supports specifying `export const components` in the MDX file to render some HTML elements as Astro components or framework components. `rehype-optimize-static` also needs to parse this JS to recognize some elements as non-static.
```

File: packages/integrations/mdx/src/README.md
Comment (bluwy) [reply packages/integrations/mdx/src/README.md:89]: I think the "emptying" part is more direct to comprehend, but the other parts looks good!
```suggestion
We can further optimize this by then also emptying the `elementStack`. This ensures that if we run this same flow for a deeper node in the tree, we don't remove the already-removed nodes from `allPossibleElements`.
```

**Comment by bluwy** [comment]
Thanks for checking some of my silly mistakes @sarah11918. Sorry you have to go through that, I should run through a grammar checker next time 😅

I've applied most of your suggestions except https://github.com/withastro/astro/pull/7221#discussion_r1207128251 that I'd like to make sure you see it before merging this.

File: packages/integrations/mdx/src/README.md
Comment (sarah11918) [reply packages/integrations/mdx/src/README.md:68]: I could be convinced to allow it, and wait until you get complaints before revisiting!  :smile:

I'll approve this, having noted the non-linear passage of time here for the record, and won't let it hold up this otherwise fabulous piece of internal documentation!  :partying_face:

**Review by sarah11918 — APPROVED** [review]

**Pull request merged**
